[RFC] Further Unify Packed and Object in TVM Runtime

tqchen opened the pull request:
This RFC proposes to further unify our PackedFunc and Object in TVM Runtime. The proposal builds on top of our past lessons and recont lessons from related project such as [matxscript](https://github.com/bytedance/matxscript)

The key improvements include: unifying type_code, solidifying AnyValue support for both stack and object values, open doors for small-string and NLP-preprocessing, and enable universal container.



[rendered](https://github.com/tqchen/tvm-rfcs/blob/unify-packed/rfcs/0097-unify-packed-and-object.md)

tqchen commented:
cc @apache/tvm-committers for visibility

leandron commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+
+Additionally, Unifying FFI and Object would bring further unification and reduction in our overall code complexity while leveling up the extensibility, so it serves as a strong improvement to the overall quality of the project. 
+
+## Drawbacks
```
Can you comment on _how_ specifically the proposed change impacts the TVM runtime, e.g. is this a breaking change in some aspect?

tqchen replied:
Please see M0 section in the Phasing part. WIll also add a reference here:

- Note that this implies (by intention) change the ABI of packed func. We will update compiler/runtime to do so.
- We will introduce adapter to support the TVMArgs during the transition, but favors moving to new state to reduce overall complexity.

More details can be found in D4

leandron reviewed (commented):
Hi @tqchen et al. This change seems positive as consolidating the generic object type. I've got a few minor points and I guess my main question which asking for a slight expansion on how the change is specifically expected to impact TVM runtimes.

Mousius commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+- TVMArgs now stores ptr of AnyView, TVMArgs[i] will now return AnyView
+- Alias: TVMRetValue = AnyValue, TVMArgValue = AnyView
+
+These changes are invisible to the users as long as they use the same source library. We will immiedately gain the ability to do universal switching when defining PackedFunc.
```
```suggestion
These changes are invisible to the users as long as they use the same source library. We will immediately gain the ability to do universal switching when defining PackedFunc.
```

Mousius commented on rfcs/0097-unify-packed-and-object.md:
````diff
@@ -0,0 +1,677 @@
+                       int* out_tcode, TVMValue* out_value);
+    ```
+    
+    - This would require packing unpacking the typecode. Note that small string passing wont work because of the lack of the seq_len padding field.
````
```suggestion
    - This would require unpacking the typecode. Note that small string passing wont work because of the lack of the seq_len padding field.
```

Mousius commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+    
+    This approach combines code and value into a single entity, this would mean a change of ABI convention in generated code. This approach makes it possible to directly return a small-str without boxing it (however if it is faced in python frontend, likely we still need to box it to str).
+    
+    We will introduce adapters to support the old calling convention, which constructs TVMAnyView on the stack then pass things over in the transitioning period.
```
What's the planned naming conventions for this and how do they eventually get transitioned to just being `TVMCPackedFunc` again?

tqchen replied:
The naming convention is being as proposed TVMAnyValue. The particular signature is a C signature for the frontend API.

tqchen replied:
Added an expanded explaination

Mousius commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+    - AnyView also cost 16bytes. Use AnyPad<n> in locals when necessary
+- **D0b**: 24 bytes
+
+Both choices are likely OK and won’ have big impact. Because we support AnyPad<n> natively in locals (and can also pass AnyPad<n> around) by taking address of it, we would recommend start with D0a — this also keeps the overall call stack cost consistent with the current design.
```
```suggestion
Both choices are likely OK and won’t have big impact. Because we support AnyPad<n> natively in locals (and can also pass AnyPad<n> around) by taking address of it, we would recommend start with D0a — this also keeps the overall call stack cost consistent with the current design.
```

tqchen replied:
Thanks for the suggestions, updated accoridngly

Mousius commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+
+### D1section: Any.type_code section convention
+
+When to store non-object type code in AnyValue.type_code. We hae the following code segmentation organization 
```
```suggestion
When to store non-object type code in AnyValue.type_code. We have the following code segmentation organization 
```

tqchen replied:
Thanks for the suggestions, updated accoridngly

Mousius reviewed (changes requested):
Hi @tqchen et al,

Looks like a good unification which will hopefully solve a bunch of issues, one specific one I'd like to highlight for you is where we're getting the wrong types out of the type system in tvmc (`IntImm` is both `int` and `bool`):
https://github.com/apache/tvm/blob/main/python/tvm/driver/tvmc/target.py#L37-L41
Am I assuming correctly that by using AnyValue and allowing it to map to the correct Python type we'll be able to properly infer the values here?

I've posted a few questions alongside a general suggestion that we orient this towards improving TVM with a reference to `matxscript` kept in Prior Art for reference rather than through-out. Otherwise most of my comments are minor spelling mistakes - it might be worth running this through a spell checker?

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+
+### N0: First class stack small string and AnyValue
+
+**Lesson from matxscript:** Data preprocessing is an important part of ML pipeline. Pre-processing in NLP involves strings and containers. Additionally, when translating programs written by users (in python), there may not be sufficient type annotations. We can common get the one of the programs below
```
This particular lesson do come from matxscript more significantly :) Including the unicode example. So I think it is approperiate to cite in both places

tqchen replied:
Rephrased here

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+
+- Need a base AnyValue to support both stack values and object.
+    - This is to provide a safety net of translation.
+- The AnyValue needs to accommodate small-string(on stack) to enable fast string processing. Specifically, note that the particular example creates a `UCS4String res` for every character of the word. If we run heap allocation for each invocation, or even do reference countings, this can become expensive.
```
Yes This does generalizes to need to support stack values. 
However, that also depends on whether such scenarios are useful in the settings we consider. In this particular case. We have not yet seen needs on other stack values yet. But the design naturally generalizes to stack values

tqchen replied:
will add a note here

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+// Note that word[i] returns a UCS4String container to match python semantics 
+// Use UCS4String stores unicode in a fixed-length 4 bytes value to ease random
+// access to the elements. 
+List unicode_split(const UCS4String& word) {
```
Yes, the code below, however, was mainly used for demonstration purposes about what can happen

tqchen replied:
added a note about this

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+**Default size of AnyValue** Any variant of AnyPad can be used as default size of AnyValue. For example, we list the following design choices
+
+- **D0a** Default to AnyPad<0> aka 16 bytes. The advantage is smaller size overall in default parameter passing.
+- **D0b** Default to AnyPad<1> aka 24 bytes. According to matx’s experience, AnyPad<1> serves well for bytedance’s internal NLP processing needs. However that was also before we had the extra AnyPad proposal. It is now possible to have AnyValue default to 16 bytes, while still create AnyPad during intermediate execution.
```
I think it is useful to put the reference under the context, so we understand the design tradeoffs here

Mousius replied:
Agree to disagree 😸

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+
+## Drawbacks
+
+The design proposal would involve changes in our runtime system. This is however, a positive step to further solidy and reduce the overall amount of concepts in the code-base, further unify packed and object, simplify and solidify our implementation along side of AnyValue and Object. Finally, it also opens doors for NLP applications and cross project interactions.
```
This does affects and cover all exsting tvm runtimes. We intended to do a refactor for all of them for consistency

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+- T3: AnyPad: an any value that have larger padded size to accomodate on stack values.
+    - When the initial value defaults to null. Both AnyValue and AnyPad, can choose to fill the small_len to be the size of total bytes available. This can help us to be able to pass small string back in C API (without template), by looking at `AnyValue*` ’s small_len field to decide the maximum bytes allowed.
+
+**Discussions**  The default size of AnyValue is 16 bytes. This means that for small string, we can use extra 8 bytes to store the string part(7 bytes if we need a tail `\0`). If we go with UCS4, we can store two extra UCS4 Char without the tail `\0`. The extra space may not be sufficient for some of the small string needs (as a reference matxscript adopts extra padding of 8 bytes to accomodate s ...
```
To make sure the reader can get most of the information, we beleive it is helpful to also put these information in context of discussion, along with the related sections. This is also a common practice in scientific writings

tqchen commented on rfcs/0097-unify-packed-and-object.md:
````diff
@@ -0,0 +1,677 @@
+}
+```
+
+- Need a base AnyValue to support both stack values and object.
````
Added additional explainations

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+### N2: Further Unify POD Value, Object and AnyValue
+
+TVM currently do have an AnyValue. Specifically `TVMRetValue` is used to hold managed result for C++ PackedFunc return can serve as any value. Additionally, if the value is object. `ObjectRef` serves as a nice way that comes with various mechanisms, including structural equality hashing.
+If we create Boxed Object for each stack values, e.g. Integer to represent int. We will be able to effectively represent every value in Object as well.
```
Added a reference

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+
+- G0: Automatic switching between object focused scenario and stack-mixed that requires AnyValue.
+- G1: Enable efficient string processing, specifically small-string support for NLP use-cases.
+- G2: Enable efficient universal container (e.g. Array that stores everything).
```
The intention is to have both Array and List follow the same container, but still enables specialization

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+- **D2b:** Always unbox to the POD value if object is a boxed value.
+    - This simplifies conversion of AnyValue into POD, since there is no need to check for boxed values.
+
+We encourage D2b when possible, this is because such conversion can be simplified in assignment. It also can help to simplify compiler side logic which only looks at POD type code but cannot handle the ObjectBoxing. 
```
Updated per suggestion

tqchen commented:
Thanks @leandron @Mousius for helpful suggestions in readability, we have incorporated most of them.  We would like still keep discussions of prior work design(in this case matx) both in prior Art as well as in the section of respective context to give clarity and context for the reader (this is also a common practice in technical scientific writing). Also updated the text with more clarifications on some of the questions asked.

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+Key elements include:
+
+- There is a zero field to enable `\0` paddings for small-str
+- The category_or_small_len field is stored in the end, to accommodate the zero padding
```
updated the text here. category means whether it is a small or big string

Mousius commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,712 @@
+      void CallPacked(TVMArgs args, TVMRetValue* rv);
+      // new convention
+      void CallPacked(int num_args, AnyView* args, AnyValue* out)
+    }    
```
Some rogue whitespace here.
```suggestion
    }
```

Mousius commented on rfcs/0097-unify-packed-and-object.md:
````diff
@@ -0,0 +1,712 @@
+    We will introduce adapters to support the old calling convention, which constructs TVMAnyView on the stack then pass things over in the transitioning period. PackedFunc will continue to support the existing TVMArgs and TVMRetValue signature, which adapts 
+    to the new calling convention.
+    ```c++
+    class PackedFunc {
````
This works for C++, what about the C runtime and `TVMFuncCall` (https://github.com/apache/tvm/blob/main/include/tvm/runtime/c_runtime_api.h#L323-L324)?

tqchen replied:
Great question. Updated the text to contain this part.

 Transition of `TVMFuncCall` also happens in two steps.
    - First step the frontend facing APIs such as `TVMFuncCall` will be kept the same, providing an adapter to call into
      the new convention under the hood.
    - Then we will update the frontend implementation, compiler, and runtime to match the new proposed convention.

Mousius commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,712 @@
+- T3: AnyPad: an any value that have larger padded size to accomodate on stack values.
+    - When the initial value defaults to null. Both AnyValue and AnyPad, can choose to fill the small_len to be the size of total bytes available. This can help us to be able to pass small string back in C API (without template), by looking at `AnyValue*` ’s small_len field to decide the maximum bytes allowed.
+
+**Discussions**  The default size of AnyValue is 16 bytes. This means that for small string, we can use extra 8 bytes to store the string part(7 bytes if we need a tail `\0`). If we go with UCS4, we can store two extra UCS4 Char without the tail `\0`. The extra space may not be sufficient for some of the small string needs (as a reference matxscript adopts extra padding of 8 bytes to accommodate  ...
```
The future reference is unnecessary to motivate the current work and you've copied it to Future Possibilities, therefore I still advocate for removing it here.

tqchen replied:
I removed that per suggestion

Mousius commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,712 @@
+
+- G0: Automatic switching between object focused scenario and stack-mixed that requires AnyValue.
+- G1: Enable efficient string processing, specifically small-string support for NLP use-cases.
+- G2: Enable efficient universal container (e.g common container for List/Array that stores everything).
```
It would be better to standardise on the container name rather than duplicate it and use up both names?

tqchen replied:
This is mainly considering the state of transition. Array currently in TVM is immutable(closer to tuple), where as List is muttable. So likely we will need to start with two container names (that shares the same underlying object). Then we can consider aliasing, consolidation or keep two versions based on the need.

tqchen commented:
@Mousius to specifically answer your questions on IntImm for bool and int. Likely this is due to the way we use container and checkers. Right now IntImm is indeed designed to hold for int and bool. And we will need to specific type checking on dtype which might be missed in such a case. Differentiating bool and int in the special section will likely resolve the issue if we believe it is important to have special casing for bool

tqchen commented on rfcs/0097-unify-packed-and-object.md:
```diff
@@ -0,0 +1,677 @@
+
+## Drawbacks
+
+The design proposal would involve changes in our runtime system. This is however, a positive step to further solidy and reduce the overall amount of concepts in the code-base, further unify packed and object, simplify and solidify our implementation along side of AnyValue and Object. Finally, it also opens doors for NLP applications and cross project interactions.
```
updated

tqchen commented:
@leandron @Mousius please take another look, i believe all the outstanding comments are addressed

cyx-6 reviewed (approved):
LGTM! We are so excited to embrace a more unified and robust runtime object system in TVM. I believe this system will not only help address some current issues, but also be a solid foundation of our TVM Unity. And hope to have some contribution with our new runtime object system implementation. :)

tqchen commented:
ping @Mousius @leandron

Mousius commented:
@tqchen Ack. I'll take another look this week.

Hzfengsy commented:
The comments so far seem have been addressed, would love to see if we have additional comments, and move forward on this

Hzfengsy commented:
Thanks for everyone's input. We are going to merge in 24 hours if there are no additional comments.

Hzfengsy commented:
This RFC is merged now. Thanks @tqchen for the proposal and the reviews from @leandron @Mousius @cyx-6
